### PR TITLE
Monotonic ordering of skin, top surface and ironing

### PR DESCRIPTION
**Pull request by Ghostkeeper**
This adds a number of settings that allow the skin, top surface and ironing to be printed in a monotonic order.

For a more thorough explanation, see the CuraEngine sister-PR. Please contain the discussion in that PR as well.

https://github.com/Ultimaker/CuraEngine/pull/1476

Contributes to CURA-7928. Implements https://github.com/Ultimaker/Cura/issues/8909.

### CODE REVIEW

**Comment by fieldOfView** [comment]
Would it be hard to implement this for the bottom too? I think many people will want their bottoms clean too...

**Comment by Ghostkeeper** [comment]
Best keep the discussions confined to the CuraEngine ticket! But this currently applies to all skin, top surface and ironing, which includes the bottom skin, yes.

**Review by rburema — APPROVED** [review]
(Also see the engine review of course! There I _do_ request a number of minor changes.)

Just to check: There is no way `infill_enable_travel_optimization` is going to be 'on' for the same sort of lines that monotonic is going to be 'true' for right? Otherwise we should prevent it here in the `enabled`s.

File: resources/definitions/fdmprinter.def.json [hunk 1]
```diff
@@ -1519,6 +1519,16 @@
                     "limit_to_extruder": "top_bottom_extruder_nr",
                     "settable_per_mesh": true
                 },
+                "skin_monotonic":
+                {
+                    "label": "Monotonic Top/Bottom Order",
+                    "description": "Print top/bottom lines in an ordering that causes them to always overlap with adjacent lines in a single direction. This takes slightly more time to print, but makes flat surfaces look more consistent.",
+                    "type": "bool",
+                    "default_value": false,
+                    "enabled": "(top_layers > 0 or bottom_layers > 0) and top_bottom_pattern != 'concentric'",
```
Comment (nallath) [inline resources/definitions/fdmprinter.def.json:1528]:
```suggestion
                    "enabled": "(top_layers > 0 or bottom_layers > 0) and (top_bottom_pattern != 'concentric' or top_bottom_pattern_0 != 'concentric')",
```
If the top_bottom_pattern_0 is not concentric, but the other pattern is, I'd argue that this setting should still be active, right?

**Comment by Ghostkeeper** [comment]
> Just to check: There is no way infill_enable_travel_optimization is going to be 'on' for the same sort of lines that monotonic is going to be 'true' for right? Otherwise we should prevent it here in the enableds.

Currently `infill_enable_travel_optimization` only applies to infill, and these settings don't apply to infill. If there is ever a skin-version of that setting, we should make that setting invisible if monotonic is enabled since currently monotonic is not optimising travel moves at all (using combing or otherwise). If we implement monotonic to optimise which monotonic segment is closest, we can re-enable the travel optimisation again also for monotonic.

**Review by nallath — APPROVED** [review]

**Comment by rburema** [comment]
> Currently `infill_enable_travel_optimization` only applies to infill, and these settings don't apply to infill.

Good, just as I thought, then.

**Pull request merged**